packages/open-next/src/adapters/image-optimization-adapter.ts

```diff
@@ -126,6 +136,63 @@ export async function defaultHandler(
 // Helper functions //
 //////////////////////
 
+/**
+ * Classifies an error and converts it to the appropriate OpenNext error type
+ * with the correct status code and logging level.
+ */
+function classifyError(e: any): IgnorableError | RecoverableError | FatalError {
+  // Default values
+  let statusCode = 500;
+  const message = e?.message || "Internal server error";
+
+  // If it's already an OpenNext error, return it directly
+  if (e && typeof e === "object" && "__openNextInternal" in e) {
+    return e;
+  }
+
+  // Determine if this is a client error (4xx) or server error (5xx)
+  const isClientError =
+    e &&
+    typeof e === "object" &&
+    (("statusCode" in e &&
+      typeof e.statusCode === "number" &&
+      e.statusCode >= 400 &&
+      e.statusCode < 500) ||
+      e.code === "ENOTFOUND" ||
+      e.code === "ECONNREFUSED" ||
+      (e.message &&
+        (e.message.includes("403") ||
+          e.message.includes("404") ||
+          e.message.includes("Access Denied") ||
+          e.message.includes("Not Found"))));
+
+  // Determine appropriate status code based on error type
+  if (e && typeof e === "object") {
+    if ("statusCode" in e && typeof e.statusCode === "number") {
+      statusCode = e.statusCode;
+    } else if ("code" in e) {
+      const code = e.code as string;
+      if (code === "ENOTFOUND" || code === "ECONNREFUSED") {
```

Comment:
Isn't `ECONNREFUSED` usually a 500 level error?

Reply:
I think an important question might be to what extent does the native NextJS Image Optimizer disregard the error codes and message it gets back and simply responds to the client in that generic way.

Then perhaps all we need to do in the adapter is ensure we're not throwing or console.error on anything ... while still returning the accurate error messages? 

This would change how this code works now.... as its trying to modify the error codes and messages to lower 5xx to 4xx so as not to crash / throw the server due to a client bad request (url)

Reply:
gotcha. Very weird that nextjs decided to swallow the 500s into 400. Ooof

Reply:
Next.js's approach of normalizing upstream 5xx errors to 4xx responses is actually quite sensible for this use case. Here's why:

**Monitoring clarity:** By converting upstream 5xx errors to 4xx, Next.js ensures that your monitoring systems only alert on problems you can actually fix, not issues with third-party services.

**Semantic correctness:** From Next.js's perspective, if a user provides a URL that leads to a broken resource, that's fundamentally a client input problem - the client provided a "bad" URL, even if it's "bad" because the upstream server failed.

**Practical operations:** Development and operations teams need clean separation between "our systems are broken" vs. "external dependencies are broken." The conversion to 4xx helps maintain this boundary.

Reply:
`Monitoring clarity` - I think you can argue the opposite too. If Nextjs converts the specific error to a generic, it's hard for you to determine what the problem is.

`Semantic correctness` - Most of the time, the "end user" is not the consumer of the image optimization request - the app is. This is why the `remotePatterns` exists. Your app service is not a service for people to arbitrarily use to optimize their images.

```
Practical operations: Development and operations teams need clean separation between "our systems are broken" vs. "external dependencies are broken." The conversion to 4xx helps maintain this boundary.
```
I'm confused by this statement, it sounds contradicting? If the error is swallowed, how would you know if it's "our system" or "their system" that's broken?